### PR TITLE
Proposal: Go Compile Time Instrumentation SIG

### PR DESCRIPTION
**Pull request by ralf0131**
This is a proposal to establish the Go compile time instrumentation SIG. The background has been discussed in https://github.com/open-telemetry/community/issues/1961. A consensus has been reached among GC/TC members, Go maintainers and multiple interested vendors.
We believe that the Go Compile Time Instrumentation SIG ensures that Go applications benefit from standardized, vendor-neutral, high-quality observability solutions that are both robust and easy to implement. Having one single, standard tool removes decision points from prospective developers, which makes the path to observability shorter & easier.

### CODE REVIEW

File: projects/go-compile-instrumentation.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+- A flexible and extensible instrumentation framework for Go at compile time
+- Out-of-box instrumentation for common libraries and frameworks in Go application
+- Comprehensive documentation that covers a wide array of topics, including how to use the framework, configuration options, advanced usage scenarios, and answers to FAQs
+
+## Staffing/Help Wanted
+The following vendors are interested in improving this area:
+
+- Alibaba
+- Datadog
+- QuesmaOrg
```
[the remark this answers is not in the document]
Comment (ralf0131) [reply projects/go-compile-instrumentation.md:33]: Fixed.

File: projects/go-compile-instrumentation.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+We also welcome everyone interested in this area to participate in the discussion.
+
+### Required staffing
+
+- Project Lead: @ralf0131(Alibaba Cloud), @dineshg13(Datadog), @pdelewski(QuesmaOrg)
+- Sponsoring GC Members: @jpkrohling
+- Seed Approvers: @yiyang0(Alibaba Cloud), @123liuziming(Alibaba Cloud), @RomainMuller(Datadog)
```
Comment (mx-psi) [inline projects/go-compile-instrumentation.md:41]: Two of these people are not OpenTelemetry Github org members. Can we reword this to say something like the following:

```suggestion

Additionally, the following people will participate in the SIG and be added as approvers once they are OpenTelemetry Github org members:

- Future Approvers: @yiyang0(Alibaba Cloud), @123liuziming(Alibaba Cloud), @RomainMuller(Datadog)
```

Comment (ralf0131) [reply projects/go-compile-instrumentation.md:41]: Looks good to me, I will accept the change.

**Review by mx-psi — APPROVED** [review]

**Comment by 123liuziming** [comment]
I approve!

**Comment by NameHaibinZhang** [comment]
I approve!

**Comment by y1yang0** [comment]
Vote: yes

**Comment by trask** [comment]
@123liuziming @NameHaibinZhang @y1yang0 can you please review the changes and approve it if it looks good to you (instead of only leaving a comment)? thanks

**Review by NameHaibinZhang — APPROVED** [review]

**Comment by 123liuziming** [comment]
> @123liuziming @NameHaibinZhang @y1yang0 can you please review the changes and approve it if it looks good to you (instead of only leaving a comment)? thanks

Sorry, I do not realize that I have the access to approve, thanks!

**Review by Cirilla-zmh — APPROVED** [review]
LGTM

**Review by y1yang0 — APPROVED** [review]

**Review by RomainMuller — APPROVED** [review]

**Review by felixge — APPROVED** [review]

**Review by svrnm — APPROVED** [review]

File: projects/go-compile-instrumentation.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+- Ease of Use: Developers can enable instrumentation with simple/zero configuration changes without manual code modifications.
+- Extensibility: The instrumentation framework can be extended to support new libraries and frameworks as they emerge in the Go ecosystem, ensuring ongoing compatibility with the evolving OpenTelemetry landscape.
+- Flexibility: Developers are enabled to control their tracing experience (i.e, specifying custom span tags in certain places; opting certain code paths or instances out of tracing, etc...), which can be delivered via a declarative/programmable scheme.
+
+By closely aligning with OpenTelemetry, the Go Compile Instrumentation SIG ensures that Go applications benefit from standardized, vendor-neutral, high-quality observability solutions that are both robust and easy to implement. Having one single, standard tool removes decision points from prospective developers, which makes the path to observability shorter & easier.
+
+## Deliverables
+
+- A flexible and extensible instrumentation framework for Go at compile time
+- Out-of-box instrumentation for common libraries and frameworks in Go application
```
Comment (jpkrohling) [inline projects/go-compile-instrumentation.md:25]: do you know already which ones will be done first?

File: projects/go-compile-instrumentation.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+### Required staffing
+
+- Project Lead: @ralf0131(Alibaba Cloud), @dineshg13(Datadog), @pdelewski(QuesmaOrg)
+- Sponsoring GC Members: @jpkrohling
+
+Additionally, the following people will participate in the SIG and be added as approvers once they are OpenTelemetry Github org members:
+
+- Future Approvers: @yiyang0(Alibaba Cloud), @123liuziming(Alibaba Cloud), @RomainMuller(Datadog)
+
+## Meeting Times
```
Comment (jpkrohling) [inline projects/go-compile-instrumentation.md:46]: do you think it's feasible to start already with rotating meeting times like the collector has? this way, most people will be able to attend 2/3 of the sessions

**Review by jpkrohling — APPROVED** [review]
my comment/suggestion isn't blocking, I'm happy to sponsor this SIG!

**Review by trask — APPROVED** [review]

**Review by dineshg13 — APPROVED** [review]

File: projects/go-compile-instrumentation.md (hunk 3 above)
Comment (pdelewski) [reply projects/go-compile-instrumentation.md:25]: We should deliver the framework first, as it is a prerequisite and serves as the foundation for library instrumentation.

Comment (ralf0131) [reply projects/go-compile-instrumentation.md:25]: Agreed.

File: projects/go-compile-instrumentation.md (hunk 4 above)
Comment (ralf0131) [reply projects/go-compile-instrumentation.md:46]: I am leaning towards start the first couple round of meeting, and adjust if there is anyone who is interested. Of course, other comments are welcome.

File: projects/go-compile-instrumentation.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# Go compile time instrumentation
```
Comment (danielgblanco) [inline projects/go-compile-instrumentation.md:1]: Could we rename this "Bootstrap Go compile time instrumentation SIG"? We intend projects to have a set of deliverables that allow it to have an end-state achievable in a given timeframe. This allows for better communication on progress towards certain milestones. I understand this project will be scoped to the initial implementation, and this SIG may continue to run later as a permanent SIG (or merged in with the Go SIG if appropriate).

The decision of forming this SIG as permanent or not can be deferred after the project is complete.

File: projects/go-compile-instrumentation.md (hunk 3 above)
Comment (danielgblanco) [reply projects/go-compile-instrumentation.md:25]: Related to my comment on the project title, I understand it may not be possible, but if we had a potential list of libraries we'd like to add instrumentation for, I think that would help to scope the project so that it has a clear exit criteria.

File: projects/go-compile-instrumentation.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+- Step 1: SIG Formation and Initial Setup
+- Step 2: Collaborative Architectural Design (1-2 months)
+- Step 3: Assignment of Implementation Owners
+- Step 4: Implementation and Integration (1-2 months)
+
+## Repo
+
+open-telemetry/opentelemetry-go-compile-instrumentation
+
```
Comment (danielgblanco) [inline projects/go-compile-instrumentation.md:65]: This does not have to be done now, but after the project is started, can you create a board and add it here? See info in https://github.com/open-telemetry/community/blob/main/project-template.md#project-board

File: projects/go-compile-instrumentation.md (hunk 5 above)
Comment (pdelewski) [reply projects/go-compile-instrumentation.md:1]: My view is that the framework will evolve over time, and the list of libraries will grow. Therefore, I see this as a never-ending process. Of course, this work could be continued as part of another SIG, such as GoLang: SDK. The decision to create this SIG was to have specific staffing—people who are interested in and skilled at these particular techniques. Otherwise, we could have done this work as part of the current GoLang: Automatic Instrumentation SIG, which focuses on eBPF. We discussed these ideas there (and even attempted it for some time by working on instrgen); however, the main issue was a lack of contributors. As a result, the decision was made to establish a separate SIG. Last but not least, why not (to be consistent) add the prefix 'Bootstrap' to the above-mentioned GoLang: Automatic Instrumentation and .NET: Automatic Instrumentation SIGs?

File: projects/go-compile-instrumentation.md (hunk 3 above)
Comment (pdelewski) [reply projects/go-compile-instrumentation.md:25]: See my comment below. I think that list of libraries may change (grow over time), but it's possible to define initial list.

File: projects/go-compile-instrumentation.md (hunk 6 above)
Comment (NameHaibinZhang) [reply projects/go-compile-instrumentation.md:65]: a potential list of libraries:
1. net/http
2. grpc
3. database/sql
4. gin
5. mux
6. gorm
7. go-redis
8. kafka
9. fasthttp
10. go-elasticsearch
11. mongo
12. logrus
13. slog/log
14. zerolog
15. zap
16. apache rocketmq
....
These are some of the libraries we are currently familiar with, and we will support them in the future based on their widespread use.

File: projects/go-compile-instrumentation.md (hunk 5 above)
Comment (ralf0131) [reply projects/go-compile-instrumentation.md:1]: I can get what Dan mean. In my opinion the SIG should be as a long running SIG(I do not see any reason that if will end for now), but a project is constraint and shall have a exit criteria.  Once we have separated these two concepts, what Dan suggest is to make it more clear about the scope of the specific project. I agree with Dan to add Bootstrap to the name. Once the SIG is bootstrapped and is running smoothly, this project can be closed, and the SIG will continue to run.

File: projects/go-compile-instrumentation.md (hunk 3 above)
Comment (ralf0131) [reply projects/go-compile-instrumentation.md:25]: The most important one the to deliver the instrumentation framework.
For the support for libraries and frameworks, my suggestion will be for each areas we provide support 1-2 libraries. The areas include:
* http/rpc: gin, grpc
* messaging: kafka
* Database: Mysql
* NoSQL: Redis
Of course, the names can be adjusted if there are other libraries worth mentioning.
Once we have these libraries instrumentation, the project can be ended safely. The support of other libraries will be continued in the long running SIG.

File: projects/go-compile-instrumentation.md (hunk 5 above)
Comment (danielgblanco) [reply projects/go-compile-instrumentation.md:1]: Thanks @ralf0131, you've explained better than I could :) I'd hope we can keep this project with a defined scope and then, of course, have the SIG running afterwards for as long as it's needed.

**Comment by danielgblanco** [comment]
Should this project proposal include mentions of the two donation proposals that will be considered as part of this work? The one from Alibaba in https://github.com/open-telemetry/community/issues/2344 and Datadog in https://github.com/open-telemetry/community/issues/2497

I think it's awesome that we can get the best learnings from each and provide the community with a unified approach. Would love to see it reflected here.

File: projects/go-compile-instrumentation.md (hunk 6 above)
Comment (ralf0131) [reply projects/go-compile-instrumentation.md:65]: Sure. I will put it into my todo list.

**Comment by ralf0131** [comment]
> Should this project proposal include mentions of the two donation proposals that will be considered as part of this work? The one from Alibaba in #2344 and Datadog in #2497
> 
> I think it's awesome that we can get the best learnings from each and provide the community with a unified approach. Would love to see it reflected here.

How about adding the  following content to the end of the proposal

## Reference
The project is a joint effort of donation proposal coming from Alibaba and Datadog. The proposals are listed as follows:
- https://github.com/open-telemetry/community/issues/2344
- https://github.com/open-telemetry/community/issues/2497

**Comment by danielgblanco** [comment]
I think that'd be great @ralf0131

File: projects/go-compile-instrumentation.md (hunk 3 above)
Comment (ralf0131) [reply projects/go-compile-instrumentation.md]: I have update the proposal with the suggested change.

File: projects/go-compile-instrumentation.md (hunk 5 above)
Comment (ralf0131) [reply projects/go-compile-instrumentation.md]: Updated as suggested.

**Comment by ralf0131** [comment]
> I think that'd be great @ralf0131

All the suggestion are applied. 
Unfortunately the CI fails and it looks like there is a dead link found in project/event-api.md file which is beyond the scope this proposal. Any ideas how to fix it?

**Comment by pdelewski** [comment]
> > Should this project proposal include mentions of the two donation proposals that will be considered as part of this work? The one from Alibaba in #2344 and Datadog in #2497
> > I think it's awesome that we can get the best learnings from each and provide the community with a unified approach. Would love to see it reflected here.
> 
> How about adding the following content to the end of the proposal
> 
> ## Reference
> The project is a joint effort of donation proposal coming from Alibaba and Datadog. The proposals are listed as follows:
> 
> * [[Donation Proposal]: compile time instrumentation for golang #2344](https://github.com/open-telemetry/community/issues/2344)
> * [[Donation Proposal]: DataDog/orchestrion - Compile-time instrumentation for golang #2497](https://github.com/open-telemetry/community/issues/2497)

I think, we could also mention that goal is also to replace https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrgen

**Comment by ralf0131** [comment]
> 
> I think, we could also mention that goal is also to replace https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrgen

Sure. How about this?

The project is a joint effort of donation proposal coming from Alibaba and Datadog to replace [Instrgen](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrgen). The proposals are listed as follows:
- https://github.com/open-telemetry/community/issues/2344
- https://github.com/open-telemetry/community/issues/2497

**Comment by trask** [comment]
> Unfortunately the CI fails and it looks like there is a dead link found in project/event-api.md file which is beyond the scope this proposal. Any ideas how to fix it?

fixed in #2503 and merged latest into your branch, should be fixed now

**Comment by ralf0131** [comment]
@danielgblanco @pdelewski Thanks for the comments, the proposed change has been committed. All the pending issues are cleared.

**Review by danielgblanco — APPROVED** [review]
This looks great, really looking forward for this start!

**Comment by danielgblanco** [comment]
This project now has 5 approvals from GC members, so ready to start! Merging.

File: projects/go-compile-instrumentation.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+We also welcome everyone interested in this area to participate in the discussion.
+
+### Required staffing
+
+- Project Lead: @ralf0131(Alibaba Cloud), @dineshg13(Datadog), @pdelewski(QuesmaOrg)
+- Sponsoring GC Members: @jpkrohling
+
+Additionally, the following people will participate in the SIG and be added as approvers once they are OpenTelemetry Github org members:
+
+- Future Approvers: @yiyang0(Alibaba Cloud), @123liuziming(Alibaba Cloud), @RomainMuller(Datadog)
```
Comment (steverao) [inline projects/go-compile-instrumentation.md:48]: Use `Alibaba` here? we have all used the affiliated name.

Comment (ralf0131) [reply projects/go-compile-instrumentation.md:48]: Good catch! I will update it together with the project board.
